[NNC] Add tile transformation in loopnest (fixed #52785)

huiguoo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #60601 [nnc] Add LoopNest::unsafe_fuseLoops to let users apply fusion on stmts that may violate our correctness checks
* #60569 [nnc] Add LoopNest::getLoopAt to retrieve a specified inner For-stmt
* **#57758 [NNC] Add tile transformation in loopnest (fixed #52785)**

Differential Revision: [D28260744](https://our.internmc.facebook.com/intern/diff/D28260744)

huiguoo committed:
[NNC] Add tile transformation in loopnest (fixed #52785)

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28260744).

navahgar commented on torch/csrc/jit/tensorexpr/loopnest.cpp:
```diff
@@ -1715,6 +1715,66 @@ std::vector<For*> LoopNest::reorder(
+  }
+  CHECK(yo);
+  yi = dynamic_cast<For*>(yo->body()->front());
+  CHECK(yi);
```
Is this needed because `splitWithMask` clones the statements in the body of the `For`?

huiguoo replied:
Yeah, `splitWithMask` clones the statements in the body : https://github.com/pytorch/pytorch/blob/b5b158a6c6de94dfb983b447fa33fea062358844/torch/csrc/jit/tensorexpr/loopnest.cpp#L1237
Can we do `Substitute` on the original stmt? If so, what's the main reason for cloning?

ZolotukhinM replied:
Cloning is probably a leftover from some old implementation that can be removed now.

navahgar replied:
Yeah, this cloning part should be removed.

navahgar commented on torch/csrc/jit/tensorexpr/loopnest.cpp:
```diff
@@ -1715,6 +1715,66 @@ std::vector<For*> LoopNest::reorder(
+    yi->setBody(body_new);
+  }
+
+  // Reorder the axes to be xo, yo, xi, yi
```
Can we use the `reorder` API here?

huiguoo replied:
The `reoderAxis` API does some extra work like deciding which loop is outter. Any concerns about the direct `reorder` here?

ZolotukhinM replied:
Tile is a composite transformation and should be built on top of primitive ones (like reorder). It is important to do that, because this allows us to be confident that all transformations are correct as long as a relatively small set of primitive transforms is correct and composite transforms are simply a combination of primitive ones. It makes it easier to test,  easier to fix bugs, and easier to refactor in future.

navahgar replied:
Precisely the reasons that Mikhail pointed out here. 

Note that `reorderAxis` is an API that only reorders two loops and handles conditionals as well. There is another API, `reorder`, which takes in a vector of loops and reorders them. But this expects the given loops to be perfectly nested. See: https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/tensorexpr/loopnest.h#L221

navahgar commented on torch/csrc/jit/tensorexpr/loopnest.h:
```diff
@@ -222,6 +222,39 @@ class TORCH_API LoopNest {
+  // into an outter iteration over the blocks and an inner traversal over all
+  // points in the block.
+  // Note that if x dim % x_factor or y dim % y_factor does not equal to 0, the
+  // loop body will be guarded by boundary checks.
```
Did you consider splitting such loops instead of adding guards? Does guards give a significant advantage here?

I believe the implementation would be lot simpler if we split the loops and not have guards?

huiguoo replied:
I tried to generate simpler stmts for later transformations like `reorder`. Probably we should consider leaving the decision of `splitWithTail` or `splitWithMask` to when we are ready for codegen (the lowering pass)? Otherwise, dealing with a large compound loop stmt after `splitWithTail`s is so painful. @ZolotukhinM

navahgar replied:
Actually, using guards (with `splitWithMask`) would be generating a more complicated code. I don't understand why `splitWithTail` will result in that. Can you show an example?

ZolotukhinM replied:
I mocked up a tile in python with `split_with_tail` and it seems working:
```python
import torch

class kernel_arena_scope(object):
    def __enter__(self):
        self.scope = torch._C._te.KernelScope()

    def __exit__(self, typ, val, traceback):
        self.scope = None

def tile(loopnest, loop_x, loop_y, factor_x, factor_y):
    enc_loops = loopnest.get_enclosing_loopnest(loop_y)
    assert(enc_loops[-1] == loop_x)

    yo, yi, yt = loopnest.split_with_tail(loop_y, factor_y)
    print("split y", loopnest)

    loop_x = loopnest.distribute_loop_over_inner_loops(loop_x)[0]
    print("distribute x", loopnest)

    xo, xi, xt = loopnest.split_with_tail(loop_x, factor_x)
    print("split x", loopnest)

    # We need to re-identify yo and yi because after split(x) we have two
    # versions of each: one in the x_inner and another in x_tail
    # But since we explicitly distributed x we know that yo is simply the first
    # and the only stmt in xi. Similarly, yi is the first and only stmt in yo.
    yo = xi.body().stmts()[0]
    yi = yo.body().stmts()[0]
    loopnest.reorder([xo, xi, yo, yi], [0, 2, 1, 3])
    print("reorder", loopnest)

    loopnest.simplify()
    print("simplify", loopnest)

with kernel_arena_scope():
    f32 = torch._C._te.Dtype.Float
    i32 = torch._C._te.Dtype.Int
    M = torch._C._te.VarHandle("M", i32)
    N = torch._C._te.VarHandle("N", i32)

    X = torch._C._te.Placeholder('X', f32, [M, N])
    Y = torch._C._te.Placeholder('Y', f32, [M, N])

    dim_args = [torch._C._te.DimArg(*args) for args in [(M, 'x'), (N, 'y')]]

    def compute(m, n):
        return X.load([m, n]) + Y.load([m, n])
    Z = torch._C._te.Compute('Z', dim_args, compute)
    loopnest = torch._C._te.LoopNest([Z])
    loopnest.simplify()
    print("original", loopnest)
    loops = loopnest.get_loops_for(Z)
    tile(loopnest, loops[0], loops[1], 8, 8)
    print("after tile", loopnest)
```

We still need to re-identify once, but I think it's not a very big deal (also, maybe we actually can avoid it somehow).

navahgar replied:
> # We need to re-identify yo and yi because after split(x) we have two
> # versions of each: one in the x_inner and another in x_tail

If we get rid of the `clone` in `splitWithTail`, we wouldn't have to re-identify yo and yi. They should be pointing to the ones in `x_inner` as before the split. The ones in `x_tail` should be the ones created new.

ZolotukhinM replied:
In contrast to TVM we don't really have a "scheduling part" - our schedules, or transformations, are immediate and not postponed. This is one of the design principles we've been sticking so far. The main benefit of doing so is that everything becomes very explicit - i.e. we have to be very specific about what happens in every situation. It might seem laborious, but that's why we need higher-level transforms like "tile".

It's also worth noticing that postponing the part where we generate tail loops doesn't actually makes the solution simpler - it just moves the complexity to another part. For instance, if we apply `split_with_mask`-based tile for a simple loopnest:
```
for x in 0..10
  for y in 0..10
    a[x,y] = ...
```
we would get
```
for xo in 0..3
  for yo in 0..3
    for xi in 0..4
      if xo*4+xi < 10
        for yi in 0..4
          if yo*4+yi < 10
            a[xo*4+xi,yo*4+yi] = 
```
It is not straightforward how to convert these mask-based loops to tail-based loops, because they've been reordered already (e.g. the condition in the innermost loop depends on `yi` and `yo` and the latter is an index of a grand-parent loop).

I would actually claim that solving it after everything is done would be harder, because we would not have a context of how we ended up with such indexes there - the information that we had while we were inside the tile transformation.

ZolotukhinM commented on torch/csrc/jit/tensorexpr/loopnest.cpp:
```diff
@@ -1715,6 +1715,66 @@ std::vector<For*> LoopNest::reorder(
+  For *yo, *yi;
+  splitWithMask(y, y_factor, &yo, &yi);
+  For *xo, *xi;
+  splitWithMask(x, x_factor, &xo, &xi);
```
What's the reason we're using `splitWithMask` instead of `splitWithTail`? Should it be configurable? I suspect on CPU `splitWithTail` would be better.

ZolotukhinM commented on torch/csrc/jit/tensorexpr/loopnest.cpp:
```diff
@@ -1715,6 +1715,66 @@ std::vector<For*> LoopNest::reorder(
+    Cond* body_new = nullptr;
+    if (yi->body()->nstmts() == 1) {
+      if (auto ycond = dynamic_cast<Cond*>(yi->body()->front())) {
+        // yi body is an if-stmt. Add the x boundary check into the condition of
+        // the if-stmt
```
Isn't it better to keep the `xi` check in the `xi` loop? If we move it to the `yi` loop we would be checking the same condition on each iteration of `yi` for no reason. It probably would make the reorder logic more complicated, but might be still worth doing.

huiguoo replied:
Agree. Will leave it there.

huiguoo committed:
Update on "[NNC] Add tile transformation in loopnest (fixed #52785)"


Differential Revision: [D28260744](https://our.internmc.facebook.com/intern/diff/D28260744)

[ghstack-poisoned]

huiguoo committed:
Update on "[NNC] Add tile transformation in loopnest (fixed #52785)"


Differential Revision: [D28260744](https://our.internmc.facebook.com/intern/diff/D28260744)

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28260744).

navahgar commented on torch/csrc/jit/tensorexpr/loopnest.cpp:
```diff
@@ -1911,6 +1911,58 @@ std::vector<For*> LoopNest::reorder(
+  auto loops = distributeLoop(xi);
+
+  // For {xo, xi, yo, yi}, reorder the axes to be xo, yo, xi, yi
+  xi = loops.front();
```
This is redundant I think. After distribute, the input loop should still be the first loop.

navahgar commented on torch/csrc/jit/tensorexpr/loopnest.cpp:
```diff
@@ -1911,6 +1911,58 @@ std::vector<For*> LoopNest::reorder(
+  xi = loops.front();
+  For* yo = dynamic_cast<For*>(xi->body()->stmts().front());
+  CHECK(yo);
+  reorderAxis(xi, yo);
```
Since this loopnest is expected to be perfectly-nested, you could use `reorder` API instead here.

huiguoo replied:
The loopnest is not perfectly-nested because `xo` has two nested loops inside, i.e., `{xi, yo, yi}` and `{xi, ytail}` so that's why we don't need to fuse any loops afterwards. In summary, we can either use `reorderAxis` in this way or use `reorder` in the way of 1)`distribute` `xo` over the two loops then 2) `reorder` {xo, xi, yo, yi} and finally 3) fuse the two distributed loops. Both are fine but the latter one seems a bit tedious to me.

navahgar replied:
I understand `xo` has two nested loops inside. But since you are only reordering `xi` and `yo`, and they are perfectly nested, you can still use `reorder` API. That API has simpler logic and also performs the transformation in-place. So, you can avoid looking up the loops again below.

navahgar commented on torch/csrc/jit/tensorexpr/loopnest.cpp:
```diff
@@ -1911,6 +1911,58 @@ std::vector<For*> LoopNest::reorder(
+
+  // Re-Identify loop handles of yo, xi, yi, ytail
+  For* xo = x;
+  yo = dynamic_cast<For*>(xo->body()->stmts().front());
```
If you use `reorder` API above, you don't have to look up these again here, since it performs the reorder transformation in-place.

navahgar commented on torch/csrc/jit/tensorexpr/loopnest.cpp:
```diff
@@ -1911,6 +1911,58 @@ std::vector<For*> LoopNest::reorder(
+  result.push_back(xi);
+  result.push_back(yi);
+  result.push_back(xtail);
+  result.push_back(ytail);
```
Why do we have to return all the loops here? They could be at different level in nesting. It will be hard to make sense of it at the caller.

We could either return just the outer most loop, or since the given loops are valid after transformation, we can just return a bool to say that it is successful.

navahgar replied:
As you noted, the tail loops are returned in awkward way. We can get around this by just returning a bool here to denote success. Since you are keeping this transform in-place, the higher-level API can extract the loops you need and return to the caller.

huiguoo replied:
Hmm, not for `xtail`. The user cannot extract the `xtail` loop because `xtail` is not nested inside `xo` (i.e., `x`, the only valid loop handle the user holds after `tile`). The user has to retrieve all loops for a buffer/tensor to find `xtail`.

huiguoo replied:
Or perhaps we can fuse the resulted loops in this way,
```
xo 
  yo
    xi
      yi
  ytail
    xi
xtail
```
returning `{xo, yo, xi, yi, ytail, xtail}` will make more sense: `ytail, xtail` are the root of the only two branches from the main loop and the user has all loop handles he/she is interested in.

navahgar replied:
Good point regarding `xtail`. I agree that we have to find a way to return some of the transformed loops so that the user can retrieve the rest.

But returning `{xo, yo, xi, yi, ytail, xtail}` looks confusing to me. Why are we returning in that order? Why are we returning only one `xi`? Why not return the other `xi` (the one after distribute) as well? If we return a vector like this, is it okay to return a `nullptr` for `xtail` if no tail was generated? 

If you prefer to return all the transformed loops, here is one suggestion I have: return 3 different vectors each representing a loop nest in the transformed loop.
For example, for the transformed loop:
```
xo
  yo
    xi
      yi
  xi
    ytail
xtail
```
we return 3 vectors:
  * main loop nest: `xo, yo, xi, yi`
  * ytail loop nest: `xo, xi, ytail`
  * xtail loop nest: `xtail`
These can be 3 parameters to the API. The ytail and xtail loop nest parameters will be set to `nullptr` if the corresponding tails were not generated. What do you think of this? 

@ZolotukhinM  do you have any other idea for this?

ZolotukhinM replied:
If we want to return all loops, I would prefer to return it as a vector of vectors, similar to how we return in `getAllLoopNestsFor`. That would convey the structure the transformation has made.

An alternative approach could be to put all these loops into a Block and return the block. This block would have a well-defined structure and the user would be able to easily locate necessary loops within it:
```cpp
Block* tile_stmt = tile(loop, 16, 16);
xo = tile_stmt.stmts()[0];
if (tile_stmt.stmts().size() > 1) 
  xtail = tile_stmt.stmts()[1];
else
  xtail = nullptr;
yo = ...;
```
This code of course could be wrapped in a nicer looking API if needed.

The drawback of this approach is that it is not used anywhere else and can be surprising.

navahgar replied:
> If we want to return all loops, I would prefer to return it as a vector of vectors, similar to how we return in getAllLoopNestsFor. That would convey the structure the transformation has made.

The reason I wasn't suggesting this is because we want to have all our transforms return a `bool`. If it is going to be an out parameter, why not have 3 out-parameters :)

navahgar replied:
> An alternative approach could be to put all these loops into a Block and return the block.

The loops will already be part of a parent block. So, we can't add them to a new block. If we return the parent block, it is not clear how to identify `xtail`.

ZolotukhinM replied:
> The loops will already be part of a parent block. So, we can't add them to a new block. If we return the parent block, it is not clear how to identify xtail.

Yes, what I meant is to construct a nested Block just for the statements from the tile transform - that block would be a statement in the initial loop's parent block. IRSimplify would get rid of that block once applied, but before that we can use it to identify loops.

huiguoo replied:
> The reason I wasn't suggesting this is because we want to have all our transforms return a bool.

Why? Are we going to have `splitWithTail` not returning `xi, xtail` but a bool? @navahgar

navahgar replied:
`splitWithTail` returns `xi` and `xtail` as out-parameters. That will stay the same. But all transformations should return a bool indicating success. See: https://github.com/pytorch/pytorch/issues/56357

huiguoo replied:
> An alternative approach could be to put all these loops into a Block and return the block. This block would have a well-defined structure and the user would be able to easily locate necessary loops within it

I am thinking of a similar approach. While returning the main loop `xo, yo, xi, yi` and the roots of tailing loops `ytail, xtail` in `tile`, we also provide another API, say `getLoopsIn()` which returns loop vectors in a root loop stmt. The users are able to retrieve all loop vectors in a tailing loop if they'd like to perform more transformations on them. Otherwise, they already have all loop handles in the main loop.

huiguoo replied:
>But returning {xo, yo, xi, yi, ytail, xtail} looks confusing to me. Why are we returning in that order? 

It matches where they appear in the transformed loop:
```
xo
  yo
    xi
      yi
  ytail
xtail
```  
`xtail, ytail` will be `nullptr` if there is no such corresponding tailing loop.

navahgar replied:
Why are we not returning the other `xi` under `ytail`?

huiguoo replied:
We are returning `ytail` and the user can use another API `getLoopIn()` to retrieve `xi`. Same for `xtail`, below is the structure of `xtail`:
```
xtail
  yo
    yi
  ytail
```
The user is able to retrieve all loop vectors in `xtail` by passing `xtail` handle to `getLoopsIn()`.

This way we speedup the path to main loop transformation as the user holds all loop handles in the main loop; but also support transformations on tail loops as we return their roots and provide an API to retrieve all loops in them.

navahgar replied:
Don't we generate the following structure?
```
xo
  yo
    xi
      yi 
  xi 
    ytail
xtail
```
i.e., `ytail` inside `xi` and not the other way around.

Given this and the logic that we should be returning the start of each loop, shouldn't we return `xi` instead of `ytail`?

huiguoo replied:
> splitWithTail returns xi and xtail as out-parameters. That will stay the same. 

`tile` transformation seems similar to `split` transformation to me. Both generates new loop axes and should return them to the user. Also, we would never return `Flase` if we return a bool in `tile`: say `x, y` are not perfectly nested, we should throw an exception because that's a pre-condition to tile; and tile two perfectly nested loops would never fail. If the specified factors are both larger than their loop bounds (a corner case I should add in the tests :smile: ), we would not touch the original stmt. it, however, an be considered as the simplified stmt for a tiled version.

huiguoo replied:
> Given this and the logic that we should be returning the start of each loop, shouldn't we return xi instead of ytail?

Ah, that's what I propose to change too : we reorder `xi` and `ytail` so the new order `ytail, xi` kinda matches the main loop `yo xi`; and this way `ytail` becomes the root of the tail loop.

huiguoo replied:
In summary, my solution is to return `xo, yo, xi, yi, ytail, xtail` as in the following structure:
```
xo
  yo
    xi
      yi
  ytail
    ytail-xi
xtail
  xtail-yo
    xtail-yi
  xtail_ytail
```
, and provide an API `getLoopsIn()` to retrieve all loop vectors in a root loop (for further transformations on tail loops `xtail, ytail` if any).

navahgar replied:
> say x, y are not perfectly nested, we should throw an exception because that's a pre-condition to tile;

We want to avoid throwing exceptions in such cases since that is still a valid IR, just an invalid input to this transformation. It is difficult when exceptions are thrown because not all callers catch exceptions, so, the process gets killed completely. Hence this issue: #56357

navahgar commented on torch/csrc/jit/tensorexpr/loopnest.cpp:
```diff
@@ -1911,6 +1911,58 @@ std::vector<For*> LoopNest::reorder(
+    ytail = dynamic_cast<For*>(ytail->body()->stmts().front());
+    CHECK(ytail);
+  }
+
```
Don't you have to fuse the loops, since you distribute above? The documentation in the `.h` file says that the output will be fused.

huiguoo replied:
It's already fused and the structure is like the following,
```
xo
  yo
    xi
      yi
  xi
    ytail
xtail
    
```

navahgar replied:
I was referring to the distribute above. But I get your point now. Since you are only distributing `xi` and that is reordered with `yo`, there is nothing to fuse here.

navahgar commented:
Thanks for the those changes. This looks much better. Some more comments inline.

huiguoo committed:
Update on "[NNC] Add tile transformation in loopnest (fixed #52785)"


Differential Revision: [D28260744](https://our.internmc.facebook.com/intern/diff/D28260744)

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28260744).

navahgar reviewed (approved):
LGTM

huiguoo committed:
Update on "[NNC] Add tile transformation in loopnest (fixed #52785)"


Differential Revision: [D28260744](https://our.internmc.facebook.com/intern/diff/D28260744)

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28260744).

huiguoo committed:
Update on "[NNC] Add tile transformation in loopnest (fixed #52785)"


Differential Revision: [D28260744](https://our.internmc.facebook.com/intern/diff/D28260744)

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28260744).
